### PR TITLE
Add support for QUIC v2

### PR DESCRIPTION
**Pull request by awelzel**

### CODE REVIEW

**Comment by awelzel** [comment]
@bbannier , any chance you might look this over? Not sure the C++ refactor makes it much better, but hope that with more versions more streamlining happens...

**Review by bbannier — APPROVED** [review]
LGTM.

I am absolutely no fan of the static variables since they make it very hard to follow what is going on, but they seem safe right now and we seem to need them since otherwise getting the right data to the calls in the new class hierarchy is pretty tedious.

As an exercise I tested whether changing all the static `std::vector` to `std::array` made any difference (also required making `HkdfCtxParam` templated over the array size, and changing `QuicPacketProtection` a variadic function to allow different param types and inside handling a la https://stackoverflow.com/questions/7230621/how-can-i-iterate-over-a-packed-variadic-template-argument-list/60136761#60136761 since the types become heterogenous and a loop cannot work). That made no measurable performance difference, likely since this code is already executed as little as possible.
